beps/0005-split-backend-discovery/README.md

```diff
@@ -0,0 +1,167 @@
+---
+title: Discovery API - Split Backends
+status: provisional
+authors:
+  - '@aramissennyeydd'
+owners:
+project-areas:
+  - core
+creation-date: 2024-02-16
+---
+
+<!--
+**Note:** When your BEP is complete, all these pre-existing comments should be removed
+
+When editing BEPs, aim for tightly-scoped, single-topic PRs to keep discussions focused. If you disagree with what is already in a document, open a new PR with suggested changes.
+-->
+
+# BEP: Discovery API - Split Backends
+
+<!-- Before merging the initial BEP PR, create a feature issue and update the below link. You can wait with this step until the BEP is ready to be merged. -->
+
+[**Discussion Issue**](https://github.com/backstage/backstage/issues/NNNNN)
+
+- [Summary](#summary)
+- [Motivation](#motivation)
+  - [Goals](#goals)
+  - [Non-Goals](#non-goals)
+- [Proposal](#proposal)
+- [Design Details](#design-details)
+- [Release Plan](#release-plan)
+- [Dependencies](#dependencies)
+- [Alternatives](#alternatives)
+
+## Glossary
+
+### Gateway Node
+
+A node that serves as the primary discovery point. A gateway node will have all of the information necessary to route traffic through your system.
```

Comment:
Hmm, my goal here is the first 3, ingress routing is not a goal with this BEP. Of the last 3, I agree, those don't all need to be implemented at once. I think there's a good case for the pairs of endpoint registration and plugin discovery as well as plugin discovery and liveness checks to be implemented at the same time though.
- endpoint registration + plugin discovery -> non standard API definitions. It would allow automatic registration of this section of the existing config-based discovery (https://github.com/aramissennyeydd/backstage/blob/287c75ab0eb7ce8b20818b1c3399072b1db3536e/packages/core-app-api/src/apis/implementations/DiscoveryApi/FrontendHostDiscovery.ts#L37-L38)
- plugin discovery + liveness checks -> unregistering plugins on crashes/network issues.

Reply:
I'd love to just solve plugin listing with this BEP but I don't think there's an easy way to do that without endpoint registration. A lot of this BEP comes from the implementation I did in #22671, which has a couple of design missteps that I've noticed already and probably more lurking around, happy to reconsider and simplify this BEP if possible. A lot of my updated version today is leveraging https://microservices.io/patterns/service-registry.html